src/HookCallbackRule.php

```diff
@@ -0,0 +1,218 @@
+    {
+        $acceptedArgs = 1;
+
+        if (isset($acceptedArgsParam)) {
+            $acceptedArgs = null;
+            $argumentType = $this->currentScope->getType($acceptedArgsParam->value);
+
+            if ($argumentType instanceof ConstantIntegerType) {
+                $acceptedArgs = $argumentType->getValue();
+            }
+        }
+
+        return $acceptedArgs;
+    }
+
+    protected function validateParamCount(ParametersAcceptor $callbackAcceptor, ?Arg $acceptedArgsParam): void
+    {
+        $acceptedArgs = $this->getAcceptedArgs($acceptedArgsParam);
+
+        if ($acceptedArgs === null) {
+            return;
+        }
+
+        $allParameters = $callbackAcceptor->getParameters();
+        $requiredParameters = array_filter(
+            $allParameters,
+            static function (\PHPStan\Reflection\ParameterReflection $parameter): bool {
+                return ! $parameter->isOptional();
+            }
+        );
+        $maxArgs = count($allParameters);
+        $minArgs = count($requiredParameters);
+
+        if (($acceptedArgs >= $minArgs) && ($acceptedArgs <= $maxArgs)) {
+            return;
+        }
+
+        if (($acceptedArgs >= $minArgs) && $callbackAcceptor->isVariadic()) {
+            return;
+        }
+
+        if ($minArgs === 0 && $acceptedArgs === 1) {
+            return;
+        }
+
+        throw new \SzepeViktor\PHPStan\WordPress\HookCallbackException(
+            self::buildParameterCountMessage(
+                $minArgs,
+                $maxArgs,
+                $acceptedArgs,
+                $callbackAcceptor
+            )
+        );
+    }
+
+    protected static function buildParameterCountMessage(int $minArgs, int $maxArgs, int $acceptedArgs, ParametersAcceptor $callbackAcceptor): string
+    {
+        $expectedParametersMessage = $minArgs;
+
+        if ($maxArgs !== $minArgs) {
+            $expectedParametersMessage = sprintf(
+                '%1$d-%2$d',
+                $minArgs,
+                $maxArgs
+            );
+        }
+
+        $message = ($expectedParametersMessage === 1)
+            ? 'Callback expects %1$d parameter, $accepted_args is set to %2$d.'
+            : 'Callback expects %1$s parameters, $accepted_args is set to %2$d.';
+
+        if ($callbackAcceptor->isVariadic()) {
+            $message = ($minArgs === 1)
+                ? 'Callback expects at least %1$d parameter, $accepted_args is set to %2$d.'
+                : 'Callback expects at least %1$s parameters, $accepted_args is set to %2$d.';
+        }
+
+        return sprintf(
+            $message,
+            $expectedParametersMessage,
+            $acceptedArgs
+        );
+    }
+
+    protected function validateActionReturnType(ParametersAcceptor $callbackAcceptor): void
+    {
+        $acceptedType = $callbackAcceptor->getReturnType();
+        $accepted = $this->ruleLevelHelper->accepts(
+            new VoidType(),
+            $acceptedType,
+            true
```

Comment:
I really don't know much about the RuleLevelHelper yet, but I see in PHPStan internals that the `$strictTypes` arg  is normally set  to `Scope::isDeclareStrictTypes()`. Does it make sense to also not enforce strict types here, but set it to whatever the current file is using too?

Reply:
To be honest I haven't figured this out myself, hence why I went with `true`. The behaviour of PHPStan itself differs depending on whether `declare(strict_types=1)` is present in the file being analysed, but I'm not sure how to vary the behaviour of an extension based on that. I might need to ask in the discussion forums.